### PR TITLE
NSSC Solana

### PR DESCRIPTION
**Pull request by suryansh-tob**
- Create NSSC Solana 
- Add Sysvar Account Check section

### CODE REVIEW

File: not-so-smart-contracts/solana/sysvar_account_check/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+During signature verification, one of the only points of authentication is the signer being checked against the actual expected address. 
+
+One of the only points of validation of the passed in sysvar account is deserializing the instruction and verifying the data length.
+
+Consider the function load_instruction_at. This takes in the inputted sysvar account data, but on its own, fails to check if the inputted sysvar account is the actual one, making the function vulnerable to spoofing.
+In this scenario, spoofing instruction data to match a valid and authorized address would make an easy attack vector to bypass signature checks.
+
+
+## Example: Wormhole Exploit (February 2022)
```
Comment (montyly) [inline not-so-smart-contracts/solana/sysvar_account_check/README.md:16]: I think we can focus on a smaller, but direct, exploit scenario, and keep the wormhole exploit as a reference.

The reason is that while the issue is simple to understand, putting it in the context of wormhole add unnecessary complexity.

File: not-so-smart-contracts/solana/sysvar_account_check/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# Missing Sysvar Account Check
+The sysvar (system account) account is often used while validating access control for restricted functions by confirming that the inputted sysvar account by the user matches up with the expected sysvar account. Without this check in place, any user is capable of passing in their own spoofed sysvar account and in turn bypassing any further authentication associated with it, causing potentially disastrous effects.
```
Comment (montyly) [inline not-so-smart-contracts/solana/sysvar_account_check/README.md:2]: Looking at the description, and the recommendations, it's a bit unclear what we are trying to describe here. 

The recommendations are mostly about using `load_instruction_at_checked` over `load_instruction_at`, so maybe the issue here is more "risky usage of load_instruction_at_checked" ?

**Review by montyly — CHANGES_REQUESTED** [review]

File: not-so-smart-contracts/solana/sysvar_account_check/README.md (hunk 2 above)
Comment (suryansh-tob) [reply not-so-smart-contracts/solana/sysvar_account_check/README.md:2]: This issue isn’t specific to `load_instructions_at`, and the same check was added for other functions such as `load_current_index` (see: https://docs.rs/solana-program/1.9.2/solana_program/sysvar/instructions/fn.load_current_index.html)

The base issue is not properly validating whether or not an authorized sysvar account has been passed in. I've fixed the recommendations so it's more clear.

File: not-so-smart-contracts/solana/sysvar_account_check/README.md (hunk 1 above)
Comment (suryansh-tob) [reply not-so-smart-contracts/solana/sysvar_account_check/README.md:46]: I've added an example contract and changed the exploit scenario so it's easier to understand - the Wormhole exploit is separate from them for further reading.

**Review by smoelius — COMMENTED** [review]
My comments are pretty nitpicky. Overall, this is a very impressive piece of work.

File: not-so-smart-contracts/solana/sysvar_account_check/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+## Mitigation
+- Solana libraries should be running on version 1.8.1 and up
+- Use `load_instruction_at_checked` and `load_current_index_checked`
+
+Utilizing the latest Solana version and referencing checked functions, especially on sensitive parts of a contract is crucial even if potential attack vectors have been fixed post-audit. 
+Leaving the system exposed to any point of failure compromises the entire system's integrity, especially while the contracts are being constantly updated.
+
+Here is the code showing the sysvar account checks added between unchecked and checked functions:
+- [load_instruction_at vs load_instruction_at_checked](https://docs.rs/solana-program/1.13.5/src/solana_program/sysvar/instructions.rs.html#186-205)
+- [load_current_index vs load_current_index_checked](https://docs.rs/solana-program/1.13.5/src/solana_program/sysvar/instructions.rs.html#107-128)
```
[the remark this answers is not in the document]
Comment (suryansh-tob) [reply not-so-smart-contracts/solana/sysvar_account_check/README.md]: Fixed - links to each function separately

File: not-so-smart-contracts/solana/sysvar_account_check/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      &account_info.try_borrow_mut_data()?,
+);
+
+  if secp_instruction.program_id != secp256k1_program::id() {
+    return Err(InvalidSecpInstruction.into());
+  }
+
+  ...
+}
+```
```
[the remark this answers is not in the document]
Comment (suryansh-tob) [reply not-so-smart-contracts/solana/sysvar_account_check/README.md:26]: Done

**Pull request merged**